### ISSUE TITLE
[Data] A streaming-compatible implemenation of repartition_by_column

### ISSUE THREAD

**Issue by wingkitlee0** [ray-project/ray#42288]
### Description

I would like to propose a streaming version of the `repartition_by_column` function in Ray Data. In particular, there is minimal data shuffling.

Currently, I have a WIP PR that I can post in a few days. However, for simplicity, I am writing it as an all-to-all operator.

In below I present some basic idea..

## Implementation

Let's take an example of a list of list of integers. This represents the blocks obtained from a read task with arbitrary block size.

Step 0: Each of the initial blocks contains multiple group keys. They are generally the blocks within each file. The groups are not necessarily sorted but they should be continuous within a block.
```python
[[1, 1, 1, 2, 2, 2], [2, 2, 3, 3, 4, 4, 6, ], [6, 7, 7, 7, 7], ... ]
```

Step 1: Map / Split. Each block is split into multiple blocks depending on the group values. This operation is stream-able because it is a 1-to-k block mapping.
```python
[[1, 1, 1], [2, 2, 2], [2, 2], [3, 3], [4, 4], [6], [6], [7, 7, 7, 7], ...]
```

Step 2: Reduce / Merge. Neighboring blocks belonging to the same group are merged into a single block. This operation is essentially the same as `itertools.groupby`. If the group of the next block is different, the current block is ready to be passed to the next stage.
```python
[[1, 1, 1], [2, 2, 2, 2, 2], [3, 3], [4, 4], [6, 6], [7, 7, 7, 7], ...]
```

### Serial Case

To see how it can be implemented as a streaming operation, we can write all the steps as iterators:
```python
def split(block: list[int]):
    """Simple splitting with itertools. IRL, we can use numpy functions"""
    for k, g in itertools.groupby(block, lambda x: x):
        yield k, list(g)

def split_blocks(blocks: list[list[int]]):
    """"Loop over each block and split each of them. Yield the key as it is more efficient than peeking into the block"""
    for block in blocks:
        for k, g in split(block):
            yield k, g

def merge_blocks(items: Iterator[tuple[int, list[int]]]):
    """"We expect the output from the split is a tuple of key and block. 

    Here we flatten the list, but in real life, we concatenate them
    """
    for key, group in itertools.groupby(items, lambda x: x[0]):
        yield key, [x for _, b in group for x in b]
```

The `repartition_by_column` function would simply be a composition of these iterators:
```python
def repartition_by_column(blocks: list[list[int]]) -> Iterator[tuple[int, list[int]]]:
    for key, block in merge_blocks(split_blocks(blocks)):
        yield key, block
```

Each block output by the merge step is ready for the next stage (e.g., `map_batches`)

### Parallel Case

To perform `repartition_by_column` in parallel, we can split the data into `P` processes and handle the boundaries. Each proces essentially handles its own queue. In real situation, this can be one or files at a time. Using the previous example, we assume one process handles one block:

```python
process  |--------0-------|  |----------1--------|  |------2------| ...
data:   [[1, 1, 1, 2, 2, 2], [2, 2, 3, 3, 4, 4, 6], [6, 7, 7, 7, 7], ... ]
```

Step 1A is essentially the same the step 1 previously. The splitting can be done within each process/queue:

```python
process  |--------0---------|  |----------1--------------|  |--------2------| ...
data:   [[1, 1, 1], [2, 2, 2], [2, 2], [3, 3], [4, 4], [6], [6], [7, 7, 7, 7], ... ]
```

Step 1B: We need to handle boundaries before the merge step. For `i`-th process, it passes the left-most block to the `i-1`-th process. This ensures the new left-most block is always the beginning of the group.
```python
process  |--------0-----------------|  |----------1-----------|  |--------2------| ...
data:   [[1, 1, 1], [2, 2, 2], [2, 2], [3, 3], [4, 4], [6], [6], [7, 7, 7, 7], ... ]
```

Step 2: The merge step can now be performed in parallel, because the groups in each process are fully contained. 
```python
process  |--------0-----------------|  |----------1-----------|  |--------2------| ...
data:   [[1, 1, 1], [2, 2, 2, 2, 2],   [3, 3], [4, 4], [6, 6],   [7, 7, 7, 7], ... ]
```

### Time Complexity

Assume we have `N` data points with `M` groups and `P` processes. The time complexity of `repartition_by_column` is `O(N/P)` assuming `N>>P` and so the boundary data shuffling (linear in `P`) is negligible. This is because each split requires the reading the group columns (linear time or faster depending on whether they are sorted or not). Each merge also linear times w.r.t. the number of rows.

### Use case

This feature has been asked by multiple individuals on slack and here, e.g., #42228  . So probably there are other cases out there:

Large datasets are often partitioned by some column values to allow efficient reading of files. In this case, the column values are bounded in each file (i.e., each group is fully contained and continuous).

There are also applications in time series, in which multiple independent time series are bundled into one file. In this case, a column of group id is added to distinguish the content.

Currently, to process individual group or time-series in these datasets, a `groupby` operation is needed, which is an all-to-all operation and has a large memory footprint.

**Comment by bdewilde** [comment]
Excited to see what you come up with, @wingkitlee0 !

**Comment by richardliaw** [comment]
hey @wingkitlee0 -- saw that you were looking for feedback on this. I think the crux of this is that 1. you want each block to go through a `1 -> K` op, and 2. you want to be able to "complete" an output block without waiting for everyone to finish. 

However, this proposal relies on a critical property which I am not sure is guaranteed, which is: 

> The groups are not necessarily sorted but they should be continuous within a block.

(and also in your example, your blocks are sorted already)

So given this, here are my thoughts

1. I think certainly we can probably do an optimization where if the dataset is sorted, we can repartition very quickly
2. more generally, we should implement a hash-repartitioning instead of a sort-based (or sort-dependent) repartition.
